src/aleph/http/client.clj

```diff
@@ -143,21 +148,32 @@
         (cond
 
           (instance? HttpResponse msg)
-          (let [rsp msg]
-
-            (let [s (netty/buffered-source (netty/channel ctx) #(.readableBytes ^ByteBuf %) buffer-capacity)
+          (if (http/decoder-failed? msg)
+            (handle-decoder-failure ctx msg response-stream)
+            (let [rsp msg
+                  s (netty/buffered-source (netty/channel ctx) #(.readableBytes ^ByteBuf %) buffer-capacity)
                   c (d/deferred)]
+
               (reset! stream s)
               (reset! complete c)
               (s/on-closed s #(d/success! c true))
               (handle-response rsp c s)))
 
           (instance? HttpContent msg)
-          (let [content (.content ^HttpContent msg)]
-            (netty/put! (.channel ctx) @stream content)
-            (when (instance? LastHttpContent msg)
-              (d/success! @complete false)
-              (s/close! @stream)))
+          (if (http/decoder-failed? msg)
+            ;; note that we are most likely to get this when dealing
+            ;; with transfer encoding chunked
+            (if-let [s @stream]
+              (do
+                ;; flag that body was not completed succesfully
+                (d/success! @complete true)
+                (s/close! s))
+              (handle-decoder-failure ctx msg response-stream))
+            (let [content (.content ^HttpContent msg)]
+              (netty/put! (.channel ctx) @stream content)
+              (when (instance? LastHttpContent msg)
+                (d/success! @complete false)
+                (s/close! @stream))))
```

Comment:
Are the `(d/success! @complete ...)` calls backwards?

It looks like you set `complete` to true when the decoder failed, and set it to false when it succeeded.

Reply:
Hrmm, maybe not, I think I'd have to dig in to see how this is being used.

Reply:
`complete` is a little bit weird inside aleph.
It basically means __Did the HTTP request got completed before receiving the LastHtpContent?__.

`false` => everything is fine
`true`  => not completed successfully

Reply:
And then, we end up here : https://github.com/clj-commons/aleph/blob/ce7e49e3c3eb0c10eae5e01b1fc579b318198c68/src/aleph/http.clj#L386-L389

Reply:
This is unintuitive, and if you can think of a better name, I'd support it.

Reply:
Unfortunately, we are committed to that `aleph/complete` key on the response.
We are better off keeping it that way.